### PR TITLE
[openssl] Add cmake wrapper to handle OPENSSL_ROOT_DIR

### LINKED ISSUE #16709: [OpenSSL] Explicitly set OPENSSL_ROOT_DIR
**Is your feature request related to a problem? Please describe.**

We have received reports from users on macOS and on Windows that are using vcpkg to install our dependencies. One of these is OpenSSL. However it can happen that cmake doesn't link to vcpkg's OpenSSL version but uses a system version instead (as well?).
This will then create link conflicts with the version packaged in vcpkg due to symbols being defined multiple times.

We assume that this happens because we also have Qt in our dependencies which itself links to OpenSSL and maybe it links to vcpkg's OpenSSL whereas our cmake files will end up picking the system libraries.

Explicitly setting `OPENSSL_ROOT_DIR` to point to vcpkg's OpenSSL directory usually resolves the issue.

**Proposed solution**

I propose explicitly setting this variable from within cmake when the vcpkg toolchain file is used (maybe only if it was not explicitly set on the command line).

**Additional context**

It seems that performing a Debug build vs a Release build can have an impact on which library version is being chosen.

For context see: https://github.com/mumble-voip/mumble/issues/4854

### PR DESCRIPTION
**Pull request by JackBoosY**
Since `OPENSSL_ROOT_DIR` may be defined externally, add cmake wrapper to specify the value as `CURRENT_INSTALLED_DIR` to avoid `FindOpenssl.cmake` to find openssl elsewhere.

Fixes #16709.

Fix kf5holidays baseline issue caused by the parallel configure:
```
CMake Error: Could not open file for write in copy operation D:/buildtrees/kf5holidays/src/v5.81.0-8d1b566dfd.clean/.clang-format.tmp
CMake Error: : System Error: Permission denied
CMake Error at D:/installed/x86-windows/share/ECM/kde-modules/KDEClangFormat.cmake:67 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  D:/installed/x86-windows/share/ECM/kde-modules/KDEFrameworkCompilerSettings.cmake:79 (include)
  CMakeLists.txt:18 (include)
```

### CODE REVIEW

**Comment by JackBoosY** [comment]
@Krzmbrzl @SamuelMarks  Can you please test this PR?

Thanks.

**Comment by SamuelMarks** [comment]
@JackBoosY Perfect

### `src/main.c`
```c
#include <stdio.h>

#include <openssl/crypto.h>


int main(void) {
    printf("OpenSSL_version_num: %lu\n", OpenSSL_version_num());
}
```

### `CMakeLists.txt`
```cmake
cmake_minimum_required(VERSION 3.0)
project(pp VERSION 0.0.0 LANGUAGES C)

find_package(OpenSSL REQUIRED)

add_executable("${PROJECT_NAME}" "src/main.c")

target_include_directories("${PROJECT_NAME}" PRIVATE "${OpenSSL_INCLUDE_DIRS}")

target_link_libraries("${PROJECT_NAME}" PRIVATE OpenSSL::SSL OpenSSL::Crypto)
```

### `sh`
```sh
$ git clone --depth=1 https://github.com/microsoft/vcpkg && cd vcpkg
$ gh pr checkout 18042
$ ./bootstrap-vcpkg.sh
$ ./vcpkg install openssl
$ cd /<source_root>/build
$ cmake -DCMAKE_TOOLCHAIN_FILE=[path]'/vcpkg/scripts/buildsystems/vcpkg.cmake' ..
-- The C compiler identification is AppleClang 12.0.5.12050022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found OpenSSL: /tmp/vcpkg/installed/x64-osx/debug/lib/libcrypto.a (found version "1.1.1k")  
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/openssl-ver/build
$ cmake --build .
[ 50%] Building C object CMakeFiles/pp.dir/src/main.c.o
[100%] Linking C executable pp
[100%] Built target pp
$ ./pp
OpenSSL_version_num: 269488319
```

**Comment by Krzmbrzl** [comment]
It always worked on my machine. We only received reports of some users that were having issues with this.
From the changes this seems to includes the workaround that we always told them to use, so LGTM :+1:

**Comment by Neumann-A** [comment]
I wonder if:
https://github.com/microsoft/vcpkg/blob/a3db16a4475b963cacf0260068c497fb72c8f3c0/scripts/buildsystems/vcpkg.cmake#L522-L548
can be moved into the wrapper?

**Comment by SamuelMarks** [comment]
(reading on my phone) unless I missed something, OpenSSL is being installed to `C:\OpenSSL`? - That doesn't seem super standard, and sounds more like a CMake contribution so their `FindOpenSSL.cmake` knows more places to look…

**Comment by Neumann-A** [comment]
> That doesn't seem super standard, and sounds more like a CMake contribution so their FindOpenSSL.cmake knows more places to look… 

Yeah it is the upstream CMake module. 
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindOpenSSL.cmake#L167-L172

**Review by ras0219-msft — APPROVED** [review]
This LGTM, though could you consider addressing @Neumann-A's comment https://github.com/microsoft/vcpkg/pull/18042#issuecomment-846950033 ?

File: versions/o-/openssl.json [hunk 1]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "15f21e44c8c035e5e3fbd8541bea39aec016612f",
```
Comment (JackBoosY) [inline versions/o-/openssl.json:4]:
```suggestion
      "git-tree": "46cecbb0f83f916b49c930f630246d44de6aa7c3",
```

File: ports/openssl/vcpkg-cmake-wrapper.cmake [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    "C:/OpenSSL-Win64/"
+    "C:/OpenSSL-Win32/lib/VC"
+    "C:/OpenSSL-Win64/lib/VC"
+    "C:/OpenSSL-Win32/lib/VC/static"
+    "C:/OpenSSL-Win64/lib/VC/static"
+)
+
+_find_package(${ARGS})
+
+set(CMAKE_SYSTEM_IGNORE_PATH ${CMAKE_SYSTEM_IGNORE_PATH_BAK})
```
Comment (JackBoosY) [inline ports/openssl/vcpkg-cmake-wrapper.cmake:36]:
```suggestion
```

File: ports/openssl/vcpkg-cmake-wrapper.cmake [hunk 3]
```diff
@@ -0,0 +1,37 @@
+set(OPENSSL_ROOT_DIR_BAK ${OPENSSL_ROOT_DIR})
+get_filename_component(OPENSSL_ROOT_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
+get_filename_component(OPENSSL_ROOT_DIR "${OPENSSL_ROOT_DIR}" PATH)
+get_filename_component(OPENSSL_ROOT_DIR "${OPENSSL_ROOT_DIR}" PATH)
+
+file(TO_CMAKE_PATH "$ENV{PROGRAMFILES}" Z_VCPKG_PROGRAMFILES)
+set(Z_VCPKG_PROGRAMFILESX86_NAME "PROGRAMFILES(x86)")
+file(TO_CMAKE_PATH "$ENV{${Z_VCPKG_PROGRAMFILESX86_NAME}}" Z_VCPKG_PROGRAMFILESX86)
+set(CMAKE_SYSTEM_IGNORE_PATH_BAK ${CMAKE_SYSTEM_IGNORE_PATH})
```
Comment (JackBoosY) [inline ports/openssl/vcpkg-cmake-wrapper.cmake:9]:
```suggestion
```

**Review by PhoebeHui — APPROVED** [review]
LGTM

File: versions/o-/openssl.json [hunk 4]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "46cecbb0f83f916b49c930f630246d44de6aa7c3",
```
Comment (JackBoosY) [inline versions/o-/openssl.json:4]:
```suggestion
      "git-tree": "eb471f523da25c78b54ca194807a1c43d8faeecc",
```

File: ports/openssl/vcpkg.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "openssl",
   "version-string": "1.1.1k",
-  "port-version": 2,
+  "port-version": 3,
```
Comment (JackBoosY) [inline ports/openssl/vcpkg.json:4]:
```suggestion
  "port-version": 4,
```

File: versions/o-/openssl.json [hunk 6]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "eb471f523da25c78b54ca194807a1c43d8faeecc",
```
Comment (JackBoosY) [inline versions/o-/openssl.json:4]:
```suggestion
      "git-tree": "4323fd67834c43c2388b3581d4224df4ed99dacf",
```

File: ports/openssl/vcpkg.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "openssl",
   "version-string": "1.1.1k",
-  "port-version": 3,
+  "port-version": 4,
```
Comment (JackBoosY) [inline ports/openssl/vcpkg.json:4]:
```suggestion
  "port-version": 5,
```

File: versions/o-/openssl.json [hunk 8]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "4323fd67834c43c2388b3581d4224df4ed99dacf",
```
Comment (JackBoosY) [inline versions/o-/openssl.json:4]:
```suggestion
      "git-tree": "9cbfcf1a35e62330f5ebddca9d0f1ea493172ef7",
```

**Comment by JackBoosY** [comment]
@strega-nil-ms Please review and merge this PR first if it's suitable for us.

**Comment by strega-nil-ms** [comment]
@JackBoosY could you update the versions files?

**Comment by strega-nil-ms** [comment]
Thanks @JackBoosY :)

**Comment by Hoikas** [comment]
Breaks consumers of static openssl on windows.

**Pull request merged**